Examples/Algorithms/Geant4HepMC/src/EventAction.cpp

```diff
@@ -138,6 +142,39 @@ void followOutgoingParticles(HepMC3::GenEvent& event,
     followOutgoingParticles(event, particle->end_vertex(), processFilter);
   }
 }
+  
+HepMC3::GenParticlePtr findLastParticleTrivialAncestor( HepMC3::GenParticlePtr particle,const std::vector<std::string>& processFilter) {
+  auto endvtx=particle->end_vertex();
+  if (!endvtx) return particle;
+  if (endvtx->particles_in().size()!=1) return particle;
+  if (endvtx->particles_out().size()!=1) return particle;
+  if (!findAttribute(endvtx,processFilter)) return particle;
+  return  findLastParticleTrivialAncestor(endvtx->particles_out()[0], processFilter);
+}
+
+void followOutgoingParticlesTrivialAncestor(HepMC3::GenEvent& event, const std::vector<std::string>& processFilter) {
+   int nupdated=0;    
+   for (;;){
+     nupdated=0;
+     for (auto p: event.particles()) {
+       auto plast=findLastParticleTrivialAncestor(p,processFilter);  
+       if (p==plast) continue;
+       nupdated++;
+       auto vertex_to_remove=p->end_vertex();
+       auto vertex_to_add=plast->end_vertex();   
+       if (vertex_to_add) { 
+         vertex_to_add->add_particle_in(p);
+         vertex_to_add->remove_particle_in(plast);
+       }
+       if (vertex_to_remove) vertex_to_remove->remove_particle_in(p);
```

Comment:
```suggestion
       vertex_to_remove->remove_particle_in(p);
```
All vertices without outgoing particles were already removed in `EndOfEventAction`. Thus this is not needed, right?

Reply:
That is a check to prevent null pointer dereference. The check should be there regardless of the underlying logic.

Reply:
Okay but is it consistent to check for both without any `else`s? What happens if only one is not a `nullptr`?

Reply:
Right, one could add there some warning for the `bad` cases when it is expected that e.g. `vertex_to_remove` is `null`.
But the problem is that one should distinguish between the legitimate case of `vertex_to_remove` being `nullptr` for a final state stable particle and the case of unexpected `nullptr`.  And it is hard to list all possible cases when something could go wrong.  Therefore no warning.
Also, if I recall that correctly the `event.remove_vertex();` has a build-in check for the argument and should be safe to call even with `nullptr`.

Reply:
HepMC checks for `nullptr`. I'm just a bit concerned about silent failures but I cannot forsee the possible cases where and why this should fail, so I'm fine with it

Reply:
Exactly.